grpc_server.go

```diff
@@ -331,8 +337,36 @@ func (s *server) Subscribe(in *hegel.SubscribeRequest, stream hegel.Hegel_Subscr
 		}
 	}
 
+	s.subLock.RLock()
+	sub := s.subscriptions[id]
+	s.subLock.RUnlock()
+	// Disconnect previous client if a client is already connected for this hardware id
+	if sub != nil {
+		sub.cancel()
+	}
+	sub = &subscription{
+		Id:           id,
+		Ip:           ip,
+		StartedAt:    startedAt,
+		InitDuration: time.Since(startedAt),
+		cancel:       cancel,
+		updateChan:   make(chan []byte, 1),
+	}
+	s.subLock.Lock()
+	s.subscriptions[id] = sub
+	s.subLock.Unlock()
```

Comment:
suffers from TOCTOU. You're assuming after RUnlock the goroutine isn't being scheduled out and somehow a sub for this id is duplicated.
Instead you can do:
```go
	sub = &subscription{
		Id:           id,
		Ip:           ip,
		StartedAt:    startedAt,
		InitDuration: time.Since(startedAt),
		cancel:       cancel,
		updateChan:   make(chan []byte, 1),
	}
	s.subLock.Lock()
	old := s.subscriptions[id]
	s.subscriptions[id] = sub
	s.subLock.Unlock()

	// Disconnect previous client if a client is already connected for this hardware id
	if old != nil {
		old.cancel()
	}
```

Reply:
fair point, much cleaner as well